consensus/XDPoS/engines/engine_v2/utils.go

```diff
@@ -218,3 +219,40 @@ func (x *XDPoS_v2) CalculateMissingRounds(chain consensus.ChainReader, header *t
 
 	return missedRoundsMetadata, nil
 }
+
+func (x *XDPoS_v2) GetBlockByEpochNumber(chain consensus.ChainReader, targetEpochNum uint64) (*types.BlockInfo, error) {
+	currentHeader := chain.CurrentHeader()
+	epochSwitchInfo, err := x.getEpochSwitchInfo(chain, currentHeader, currentHeader.Hash())
+	if err != nil {
+		return nil, err
+	}
+	epochNum := x.config.V2.SwitchBlock.Uint64()/x.config.Epoch + uint64(epochSwitchInfo.EpochSwitchBlockInfo.Round)/x.config.Epoch
+	// since below function GetEpochSwitchInfoBetween(chain, start, end) return nil if start == end, we early return the result
+	if targetEpochNum == epochNum {
+		return epochSwitchInfo.EpochSwitchBlockInfo, nil
+	}
+	if targetEpochNum > epochNum {
+		return nil, errors.New("input epoch number > current epoch number")
+	}
+	if targetEpochNum < x.config.V2.SwitchBlock.Uint64()/x.config.Epoch {
+		return nil, errors.New("input epoch number < v2 begin epoch number")
+	}
+	epoch := big.NewInt(int64(x.config.Epoch))
+	estblockNumDiff := new(big.Int).Mul(epoch, big.NewInt(int64(epochNum-targetEpochNum)))
+	estBlockNum := new(big.Int).Sub(epochSwitchInfo.EpochSwitchBlockInfo.Number, estblockNumDiff)
+	if estBlockNum.Cmp(x.config.V2.SwitchBlock) == -1 {
+		estBlockNum.Set(x.config.V2.SwitchBlock)
+	}
+	estBlockHeader := chain.GetHeaderByNumber(estBlockNum.Uint64())
+	epochSwitchInfos, err := x.GetEpochSwitchInfoBetween(chain, estBlockHeader, currentHeader)
+	if err != nil {
+		return nil, err
+	}
+	for _, info := range epochSwitchInfos {
```

Comment:
So I guess this is linear algorithm to loop back from current epoch to the target one.